Horizontal fusion

trentlo opened the pull request:
Horizontally fuse (independent) computations after vertical fusion is done. It helps reduces kernel launch overhead while increasing kernels launch dims (and hence parallelism). I observed around 0-5% performance gains on various models.

Note that this PR depends on functionality of a previous one:
https://github.com/tensorflow/tensorflow/pull/34880

trentlo committed:
Implement horizontal fusion.

 - It reduces kernel launch overhead and increases lauch dims by horizontally
   fusing indepedent computations.

trentlo committed:
Minor cleanup for horizontal fusion.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+          if (user->opcode() == HloOpcode::kGetTupleElement) {
+            // Skip GTE.
+            return is_consumer_the_only_user(*user);
+          } else if (user == consumer) {
```
The large number of lambdas defined inside this function body isn't ideal for readability. In particular, this lambda captures `consumer` which is defined ~20 lines above and is called ~60 lines below. Please consider to extract the functions into an unnamed namespace and pass all arguments explicitly.

trentlo replied:
Done.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+      };
+
+  auto is_profitable = [](const HloInstruction& insn) -> bool {
+    // Do not fuse computation with too large shapes or too many instructions,
```
s/computation/instruction

trentlo replied:
Will do.

trentlo replied:
Done.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+  auto is_profitable = [](const HloInstruction& insn) -> bool {
+    // Do not fuse computation with too large shapes or too many instructions,
+    // because it may not be easily profitable.
+    constexpr int64 kShapeThreshold = 128 * 2048;
```
Would it make sense to define a threshold on the number of bytes rather than the number of elements?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+
+  auto is_profitable = [](const HloInstruction& insn) -> bool {
+    // Do not fuse computation with too large shapes or too many instructions,
+    // because it may not be easily profitable.
```
Can you add some information how you came up with these thresholds. It's interesting to understand how magic the magic number are for others considering to change them.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+      }
+    }
+
+    // Having too many instructions is not easily profitable.
```
Is the rationale here that the kernel may become compute bound?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+      VLOG(2) << "Reject unsupported fusion insn " << insn->ToString();
+      continue;
+    } else if (!is_consumer_the_only_user(*insn)) {
+      VLOG(2) << "Reject maybe illegal insn " << insn->ToString();
```
Can you say why it may be illegal. This is about cycles, right?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+    return absl::Span<HloInstruction*>();
+  }
+
+  // Fusing too many computations at a time may not be easily profitable and
```
Did you see FusionWouldBeTooLarge? Would this be useful here too?
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/gpu_fusible.cc#L298

trentlo replied:
I have considered it but there are some difficulties I see:
1. The horizontally fused kernel is much larger (e.g., in terms of the number of instructions) than the vertical fusions. So, they need different metrics.
2. We consider fusing many computations at a time, while `FusionWouldBeTooLarge` considers two fusion instructions at a time.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+}
+
+Status HorizontalFusionImpl::CreateFusedComputation(
+    absl::Span<HloInstruction*> fused_fusion_insns,
```
Did you try MergeFusionInstructionIntoMultiOutput and friends? I know you need additional Reshapes and Slices, but maybe it's possible to add them after the merge.

trentlo replied:
One difficulty could be that we need a mechanism to record the outputs of the fused computations (for later insertion of concat, etc). Using `MergeFusionInstructionIntoMultiOutput` may not be cleaner. Any thoughts on that?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+  }
+
+  // Sort according to the number of outputs and instruciton counts.
+  std::sort(fusion_insns_.begin(), fusion_insns_.end(),
```
Please motivate the need for sorting in the comment.

trentlo replied:
Will add some comments.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:
```diff
@@ -0,0 +1,244 @@
+namespace gpu {
+namespace {
+
+class HorizontalFusionTest : public HloTestBase {};
```
All tests are RunAndCompare integration tests. Integration tests have their spot, but it's a good idea to have unit tests too. Please test the pass you are adding in isolation at the HLO level, i.e. pass in an HLO test fixture and verify the output is the HLO you expected. Please also test cases in which you choose not to fuse, e.g. to avoid cycles.

trentlo replied:
Agreed.

trentlo replied:
Added.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,109 @@
+// gain.
+//
+// Theoretically speaking, we can (horizontally) fuse kernels as long as no
+// new cycles are created after the fusion. However, it requires a somewhat
```
Agreed.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,109 @@
+// new cycles are created after the fusion. However, it requires a somewhat
+// cumbersome cycle detection checks; also, we observe that naive horizontal
+// fusion of arbitrary kernels may not be profitable due to control divergence
+// and possible increase of memory bandwidth pressure (if some instructions are
```
Yes. Will make it more explicit.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,109 @@
+// and possible increase of memory bandwidth pressure (if some instructions are
+// not row-major). In practice, a simple yet effective heuristic is used to
+// avoid these issues while addressing the known beneficial cases. That is, we
+// simply search for fusion candidates by looking at computations whose outputs
```
Taken.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,109 @@
+//  Mul     Add
+//   |       |
+//   v       v
+//   o0      o1
```
Yes (in the current implementation).

I want to point out that it works well in practice to capture the case of training optimizers and other similar cases involving variables (e.g., computing L2Loss), as they typically are consumed by ROOT tuple.

thomasjoerg replied:
In that case the ROOT would be the common consumer, right? The example will be most helpful to readers if it is close to the current implementation.

trentlo replied:
Right. Re-wrote some parts of the comments so that the illustrative example is close to the current implementation.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,109 @@
+// Note that this style provides an important advantage that kernels of
+// different shapes can be horizontally fused. The first pair of reshapes
+// (i.e., Reshape0 and Reshape1) reshape the dims to 1 dimension, so that the
+// outputs of the fused kernels can (always) be concatenated. The second pair
```
It can work beyond element-wise instructions. I have left some more related comments in another reply.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+  } else {
+    return root->operands();
+  }
+}
```
A bit confusing but the end of the namespace is here:
https://github.com/tensorflow/tensorflow/blob/cd68827e01d454937399bafcdb1eb4b9a116678a/tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc#L428

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+  absl::flat_hash_set<HloInstruction*> fusion_insns;
+  for (auto opnd : consumer->operands()) {
+    auto predecessor = opnd->LatestNonGteAncestor();
+    if (predecessor->opcode() == HloOpcode::kFusion) {
```
The primary reason is to address the main use case in my mind, i.e., training optimizers. If we see cases where fusing unfused HLO instructions are important, we can extend the capability.

thomasjoerg replied:
To be fair, this is inline with how fusion passes in XLA have been developed so far. Other passes treat fused and unfused ops differently too.
Sometimes I wish they wouldn't though, because there is no fundamental reason or benefit in doing so. The downside is that reasoning about optimized HLO ("Why do I see the fusions I see?") gets harder.

thomasjoerg replied:
Just to be clear, I agree with you. Extending this to unfused HLOs can be done later, if needed, and is worth a separate PR.

trentlo replied:
See your point.

trentlo reviewed (commented):
Responded to the comments. Revising the codes.

trentlo committed:
Polishing coding style and comments.

trentlo committed:
Factor out lambdas in HorizontalFusionImpl.

trentlo committed:
Comment polishing.

trentlo committed:
Add some more unittests for horizontal fusion.

In addition, we record the execution time of the tests here, showing the
optimization effects of horizontal fusion, measured by --xla_hlo_profile. The
accumulated kernel execution time in GradientDescentOptimizerLike is reduced
from 2.39ms to 311us; the execution time in RMSProp is reduced from 980us to
112us.

Before horizontal fusion:
2019-12-10 22:05:45.215015: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for GradientDescentOptimizerLike: (2.39 ms @ f_nom)
2019-12-10 22:05:48.877372: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for RMSPropLike: (980 us @ f_nom)

After horizontal fusion:
2019-12-10 22:05:03.831600: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for GradientDescentOptimizerLike: (311 us @ f_nom)
2019-12-10 22:05:13.513901: I tensorflow/compiler/xla/service/executable.cc:208] Execution profile for RMSPropLike: (112 us @ f_nom)

trentlo commented on tensorflow/compiler/xla/service/gpu/gpu_compiler.cc:
```diff
@@ -265,6 +266,13 @@ Status GpuCompiler::OptimizeHloModule(
     fusion.AddPass<HloDCE>();
     TF_RETURN_IF_ERROR(fusion.Run(hlo_module).status());
+
+    HloPassFix<HloPassPipeline> horizontal_fusion("horizontal_fusion");
```
Done.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+}
+
+// Return the number of outputs of the fused computation.
+size_t GetOutputSizeOfFusion(const HloInstruction& insn) {
```
Done.

trentlo reviewed (commented):
Should have responded to all of your previous comments. Please help to take a look again when you have a moment. Thanks.

trentlo committed:
Polishing comments and coding styles.

gbaned commented:
@trentlo Can you please resolve conflicts? Thanks!

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,447 @@
+      return false;
+    }
+
+    return true;
```
> The only requirement for the computation itself should be that the fused computations have the same number of outputs.

Interesting, which non-element-wise ops that satisfy this property? I can think of bitcast, reshape, reverse, and transpose. Is that right? Do reshapes and transposes make sense for horizontal fusion?

trentlo replied:
They are all fine as long as they are in the kLoop fusion.

In other words, we rely on _all_ the concatenated outputs in the (slice) input fusion to have the same shape, so that the kernel generation can use this shape as kernel launch dim. To ensure the concatenated outputs having the same shape, we currently rely on two requirements--(a) the fused computations are kLoop fusion and (b) the fused computations have the same number of outputs.

Then, it guarantees that the following two concatenated outputs have the same shapes, assuming we have two fused kLoop computations, each having two outputs (`o1` and `o2`):
`concatenate(computation1.o1, computation2.o1)`
and `concatenate(computation1.o2, computation2.o2)`

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,470 @@
+bool IsProfitableFusionCandidate(const HloInstruction& instr) {
+  CHECK(instr.opcode() == HloOpcode::kFusion);
+  constexpr int64 kShapeThreshold = 128 * 2048;
+  constexpr int64 kInstrCountThreshold = 30;
```
How specific are these thresholds to a GPU architecture?

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,470 @@
+    }
+  }
+
+  // Sort according to the number of outputs and instruciton counts, because
```
Typo: instruction

trentlo replied:
Good catch. Fixed.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,470 @@
+    } else if (fusion_instrs_[left]->fused_instruction_count() !=
+               fusion_instrs_[right]->fused_instruction_count()) {
+      // Do not fuse computations of different instruction counts as it may
+      // introduce control divergence.
```
The same number of instructions could still result in control divergence, right? It may be worth pointing out that this is a heuristic.

trentlo replied:
Right. Revised the comments to make it clearer.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:
```diff
@@ -0,0 +1,328 @@
+ fused_computation.1 {
+   arg.1 = f16[123]{0} parameter(0)
+   arg.2 = f16[123]{0} parameter(1)
+   ROOT mul.1 = f16[123]{0} multiply(arg.1, arg.2)
```
HLO text allows comments starting with //.
Please add a comment why horizontal fusion will not happen here.

trentlo replied:
That is cool. In hlo text comments added.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:
```diff
@@ -0,0 +1,328 @@
+
+  EXPECT_TRUE(GpuHorizontalFusion().Run(module.get()).ValueOrDie());
+  EXPECT_TRUE(HloDCE().Run(module.get()).ValueOrDie());
+
```
HLO matchers provide a convenient way to capture test expectations at the HLO level. Please add such expectations to your tests. Whoever breaks one of your tests going forward will have an easier time figuring out what's wrong if test expectations are not only given RunFileCheck or RunAndCompare-style.

For example: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/service/gpu/multi_output_fusion_test.cc#L73

trentlo replied:
That is cool. Updated.

thomasjoerg reviewed (commented):
Looking really good. Just a few more comments.

trentlo committed:
Polish coding styles and comments based on review feedback.

In addition, use hlo_matcher to verify resultant DAGs instead of LLVM filecheck.

trentlo committed:
Merge branch 'upstream_master_dec19' into horizontal_fusion_github

trentlo reviewed (commented):
Thanks for the review again! Thomas.

I should have responded to all your existing comments. Please help to take a look again when you have a moment.

trentlo commented:
The error message appears not related to the PR. Could we run the regression test again?

`4 errors detected in the compilation of "T:/tmp/nvcc_inter_files_tmp_dir/tmp79b9es58/debug_ops_gpu.cu.compute_70.cpp1.ii".`

trentlo commented:
One previous failure is gone. The error messages in the remaining failure are all about download failures. See below for example. Would you mind running the regression test again? @gbaned 

```
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/google/pasta/archive/v0.1.8.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found
WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/googleapis/google-cloud-cpp/archive/v0.16.0.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 404 Not Found

```

gbaned commented:
@trentlo Can you please address Ubuntu Sanity errors? Thanks!

trentlo commented:
> @trentlo Can you please address Ubuntu Sanity errors? Thanks!

Should have been fixed now. It was a BUILD file format issue.

It appears someone has to approve it again for regression?

trentlo committed:
Fix a buildifier format issue.

sanjoy reviewed (approved):
Rubber-stamping based on Thomas's prior approval.

sanjoy reviewed (changes requested):
This PR fails with an internal model: https://gist.github.com/sanjoy/1a23cba0e7c53419349102b53d221708 with the error: "Invalid argument: Cannot concatenate arrays with different element types: F16 vs PRED".  Can you please take a look?

trentlo commented:
> This PR fails with an internal model: https://gist.github.com/sanjoy/1a23cba0e7c53419349102b53d221708 with the error: "Invalid argument: Cannot concatenate arrays with different element types: F16 vs PRED". Can you please take a look?

Thanks for the reporting. Will take a look at it soon.

trentlo commented:
> > This PR fails with an internal model: https://gist.github.com/sanjoy/1a23cba0e7c53419349102b53d221708 with the error: "Invalid argument: Cannot concatenate arrays with different element types: F16 vs PRED". Can you please take a look?
> 
> Thanks for the reporting. Will take a look at it soon.

I guess we missed a consideration that we should not horizontally fuse instructions if their output types are not the same (since concat cannot support this). Will update the PR soon.

trentlo committed:
Do not fuse fusions with different output types.

It is forbidden as the concatenate (inserted for horizontal fusion) requires
the input operands to have the same type.

trentlo committed:
Minor polishing.

trentlo commented:
> This PR fails with an internal model: https://gist.github.com/sanjoy/1a23cba0e7c53419349102b53d221708 with the error: "Invalid argument: Cannot concatenate arrays with different element types: F16 vs PRED". Can you please take a look?

Fixed.

Please help to take a look when you get a moment. Happy new year~

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -215,17 +227,25 @@ void HorizontalFusionImpl::FusionCandidates::Initialize(
+      fusion_instrs_.begin(), fusion_instrs_.end(),
+      [&](const HloInstruction* a, const HloInstruction* b) {
+        PrimitiveType output_type_a =
+            GetOutputsOfFusion(*a)[0]->shape().element_type();
```
The `[0]` is safe because `GetOutputsOfFusion` above will evaluate to `false` when it isn't. However, this is subtle and easy to break as code evolves. In fact, I looked at a very similar bug just today. Please explicitly check the assumption that fusions have a unique output type. You can introduce a helper `GetUniqueOutputType` that CHECK fails otherwise.

trentlo replied:
Fixed. It is indeed better and clearer to have `GetUniqueOutputTypeOfFusion`. Thanks for the comments.

trentlo committed:
Add a help function GetUniqueOutputTypeOfFusion.

It is safer and clearer for getting the unique output type of fusion.

trentlo committed:
Minor error message polishing.

trentlo reviewed (commented):
Updated. Please help to take a look. Thanks.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,514 @@
+  CHECK(fusion_instr.opcode() == HloOpcode::kFusion);
+  auto instrs = fusion_instr.fused_instructions_computation()->instructions();
+  for (auto instr : instrs) {
+    if (instr->shape().layout().format() != DENSE) {
```
I assume SPARSE is not really used (for GPU at least). Let me know if my understanding is not accurate.

If the format is SPARSE, we should return false too? as the backend cannot really support it?

trentlo committed:
Minor comment polishing.

trentlo commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.h:
```diff
@@ -0,0 +1,112 @@
+// No extra copies are introduced by the horizontal fusion. Besides Reshape2
+// and Reshape3, the other instructions are fused into an input fusion; the
+// output dims of the concatenate will be used as the kernel launch dims.
+// Instruction bitcasts can be used for Reshape2 and Reshape3 as long as the
```
I polished the comment to make explicit the condition when we can use bitcasts here (instead of using reshape).

If the outputs of mul and add are not row-major, we must use reshape, which will introduce copies.

trentlo reviewed (commented):
Two last comments from me. They should be minor. I just want to make things explicit as possible.

trentlo commented:
> Updated. Please help to take a look. Thanks.

Ping.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,514 @@
+      bound_operands->push_back(bound_opnd);
+    }
+  }
+  *uniq_computation = b.Build();
```
A few tests (e.g. `//tensorflow/compiler/xla/tests:prng_test_gpu`) seem to CHECK fail on this line.
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/tests/BUILD#L1751

```
F0117 05:31:52.164265    7056 hlo_computation.cc:62] Check failed: nullptr != root ((null) vs. (null)) 
*** Check failure stack trace: ***
...
xla::HloComputation::Builder::Build()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::CreateFusedComputation()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::Fuse()
xla::gpu::(anonymous namespace)::HorizontalFusionImpl::Run()
...
```

trentlo replied:
Fixed. It is a case where fusion instructions have no parameters at all. Then, no instructions was added into the HloComputation (and hence no roots) but existence of a root instruction is always required by this class.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc:
```diff
@@ -0,0 +1,514 @@
+    for (size_t j = 0; j < old_params.size(); ++j) {
+      auto bound_opnd = fused_fusion_instrs[i]->mutable_operand(j);
+      // in a form of param_i_j
+      auto new_param = b.AddInstruction(HloInstruction::CreateParameter(
```
`new_param` is unused. Can the assignment be dropped?

trentlo replied:
Fixed. The local variable can be dropped.

thomasjoerg commented on tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:
```diff
@@ -0,0 +1,367 @@
+        shape, HloOpcode::kAdd, ms_out, epsilon_broadcasted));
+    auto ms_eps_rsq = builder.AddInstruction(
+        HloInstruction::CreateUnary(shape, HloOpcode::kRsqrt, ms_eps));
+    auto grad_ms_eps_rsq = builder.AddInstruction(HloInstruction::CreateBinary(
```
`grad_ms_eps_rsq` is unused. Can the line be removed or just the local variable?

trentlo replied:
Fixed. The wiring was not as intended. My oversight.

thomasjoerg commented:
Hi @trentlo, I ran benchmarks internally. The good news is, I did not spot any regressions. Unfortunately, none of the benchmarks in the suite shows a speedup either. That's not a blocker, but ideally, we'd add one to safeguard against accidental regressions. What open-source benchmarks did you use for your own testing?

trentlo commented:
> Hi @trentlo, I ran benchmarks internally. The good news is, I did not spot any regressions. Unfortunately, none of the benchmarks in the suite shows a speedup either. That's not a blocker, but ideally, we'd add one to safeguard against accidental regressions. What open-source benchmarks did you use for your own testing?

Good to hear that there is no repression. :-)

To see performance gain, you need to enable resource variables, so that the optimizers are lowered into XLA. In TF1.x, variables are not lowered into XLA (i.e., only resource_varialbes are lowered).

You may try Resnet50 with **adam** in [tf_cnn_benchmark](https://github.com/tensorflow/benchmarks/tree/master/scripts/tf_cnn_benchmarks). I tried it with TF1.x (as it did not support TF2 few weeks ago) and see around 4% and 2% speedup with fp32 and fp16, respectively. I just saw that they updated the script and it may be compatible with TF2 now; I have never tried it with TF2 and I will find time to try it.

(BTW I will take a week off until 1/24. My response time will be longer than usual.)

thomasjoerg commented:
Welcome back, Trent. I hope you had an enjoyable vacation!

Thanks for your instructions for benchmarking. Just to make sure it you didn't miss it: Some tests run into CHECK failures. See my previous comments.

Unfortunately the open-source CI did not catch these test breakages. This is a known issue and a fix is underway.

trentlo commented:
> Welcome back, Trent. I hope you had an enjoyable vacation!
> 
> Thanks for your instructions for benchmarking. Just to make sure it you didn't miss it: Some tests run into CHECK failures. See my previous comments.
> 
> Unfortunately the open-source CI did not catch these test breakages. This is a known issue and a fix is underway.

Got it. Let me know if there is anything I can help with.

sanjoy commented:
> Got it. Let me know if there is anything I can help with.

Can you fix the `CHECK`-failures Thomas mentioned [above](https://github.com/tensorflow/tensorflow/pull/34882#discussion_r367937189)?  Or are they already fixed?

trentlo commented:
> > Got it. Let me know if there is anything I can help with.
> 
> Can you fix the `CHECK`-failures Thomas mentioned [above](https://github.com/tensorflow/tensorflow/pull/34882#discussion_r367937189)? Or are they already fixed?

No problem to provide a fix to it. However, @thomasjoerg did you mean someone already fix it and a commit is on the way?

sanjoy commented:
> No problem to provide a fix to it. However, @thomasjoerg did you mean someone already fix it and a commit is on the way?

I suspect in " This is a known issue and a fix is underway." @thomasjoerg was referring to "Unfortunately the open-source CI did not catch these test breakages.".  I'll ping him internally to reply here, but we're outside his working hours.

thomasjoerg commented:
> > No problem to provide a fix to it. However, @thomasjoerg did you mean someone already fix it and a commit is on the way?
> 
> I suspect in " This is a known issue and a fix is underway." @thomasjoerg was referring to "Unfortunately the open-source CI did not catch these test breakages.". I'll ping him internally to reply here, but we're outside his working hours.

Hi @trentlo, @sanjoy is right. The bug has slipped through CI and hence you see "All checks have passed". However, the CHECK failure is a real bug in this PR and needs to be fixed. I mentioned the build target above. You should be able to reproduce it.

trentlo commented:
> Hi @trentlo, @sanjoy is right. The bug has slipped through CI and hence you see "All checks have passed". However, the CHECK failure is a real bug in this PR and needs to be fixed. I mentioned the build target above. You should be able to reproduce it.

Ah. I was confused. Looking into it then.

trentlo committed:
Always gives a dummy root instruction when building a computation.

trentlo committed:
Removed unused local variable.

trentlo committed:
Fix wrong wiring in a test case.

trentlo committed:
Explicitly replace dummy root instruction.

trentlo reviewed (commented):
Fixed. Please help to take a look when you have a moment. Thanks.

trentlo commented:
There is a new build error probably because someone changed the build flag. Fixing it.

```
ERROR: /Volumes/BuildData/tmpfs/src/github/tensorflow/tensorflow/compiler/xla/service/gpu/BUILD:1696:1: Couldn't build file tensorflow/compiler/xla/service/gpu/_objs/horizontal_fusion_test/horizontal_fusion_test.o: C++ compilation of rule '//tensorflow/compiler/xla/service/gpu:horizontal_fusion_test' failed (Exit 1)
tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:205:45: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long long' in initializer list [-Wc++11-narrowing]
    auto shape = ShapeUtil::MakeShape(F32, {i + 1, 1024});
                                            ^~~~~
tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:205:45: note: insert an explicit cast to silence this issue
    auto shape = ShapeUtil::MakeShape(F32, {i + 1, 1024});
                                            ^~~~~
                                            static_cast<long long>( )
tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:295:48: error: non-constant-expression cannot be narrowed from type 'unsigned long' to 'long long' in initializer list [-Wc++11-narrowing]
    auto shape = ShapeUtil::MakeShape(F32, {2, 1024 + i});
                                               ^~~~~~~~
tensorflow/compiler/xla/service/gpu/horizontal_fusion_test.cc:295:48: note: insert an explicit cast to silence this issue
    auto shape = ShapeUtil::MakeShape(F32, {2, 1024 + i});
                                               ^~~~~~~~
                                               static_cast<long long>( )
2 errors generated.
```

trentlo committed:
Fixed a build error due to -Wc++11-narrowing.

trentlo commented:
> There is a new build error probably because someone changed the build flag. Fixing it.
> 

It is because of some narrowing checks specific to Clang. Should have been fixed. (There was another mlir test error; I assume it is not related to this commit.)

@thomasjoerg, I will need to bother you to kick off the CI again.

sanjoy commented:
> @thomasjoerg, I will need to bother you to kick off the CI again.

Done.

trentlo commented:
> > @thomasjoerg, I will need to bother you to kick off the CI again.
> 
> Done.

Thank you!

sanjoy reviewed (approved):
lgtm
